feat(core models): Impelement the core models

abdelsalam101 opened the pull request:
Impelemented the following models:
- Created a custom `User` model using AbstractBaseUser and PermissionsMixin
- Added roles: customer, seller, and shipping company
- Implemented CustomUserManager with proper `create_user` and `create_superuser`
- Added related models: Address, CustomerProfile, SellerAccount, and ShippingCompany
- Linked `Product` to `SellerAccount`

Closes #6
## 📌 Summary

Implemented a custom user model with role-based architecture to support multiple user types (Customer, Seller, Shipping Company). This was needed to allow scalable user management and authentication logic for Dropz's core functionality.

## ✅ What Was Done

- [x] Created a `User` model inheriting from `AbstractBaseUser` and `PermissionsMixin`
- [x] Added custom user manager (`CustomUserManager`) with `create_user` and `create_superuser`
- [x] Added `role` field to support `customer`, `seller`, `shipping_company`
- [x] Created supporting models:  
  - `Address` (linked to user)  
  - `CustomerProfile` (linked one-to-one to user)  
  - `SellerAccount` (linked one-to-one to user)  
  - `ShippingCompany` (linked one-to-one to user)
- [x] Linked `Product` to `SellerAccount` for seller-based ownership

## 🧪 How to Test

1. Run `docker compose up`
2. Go to `http://localhost:8000/admin`
3. Create a new user with a role (`customer`, `seller`, or `shipping_company`)
4. Login using the new user's email and password
5. Test relationships:
   - Create related profile (e.g., `CustomerProfile`, `SellerAccount`)
   - Add product for a seller
6. Run tests (if any) with `docker compose exec web pytest`

Closes #6

## 📋 Checklist Before Merge

- [x] Code is self-documented and clean
- [ ] Tests are added/updated
- [ ] Docs/README updated if necessary
- [x] I rebased onto latest `dev` and resolved any conflicts

abdelsalam101 committed:
feat(core models): Impelement the core models

Impelemented the following models:
- Created a custom `User` model using AbstractBaseUser and
  PermissionsMixin
- Added roles: customer, seller, and shipping company
- Implemented CustomUserManager with proper `create_user` and
  `create_superuser`
- Added related models: Address, CustomerProfile, SellerAccount, and
    ShippingCompany
- Linked `Product` to `SellerAccount`

Closes #6

AAEmara commented on backend/accounts/models.py:
```diff
@@ -1,3 +1,103 @@
+    def create_superuser(self, email, password=None, **extra_fields):
+        extra_fields.setdefault('is_staff', True)
+        extra_fields.setdefault('is_superuser', True)
+        return self.create_user(email, password, **extra_fields)
```
Enforce email and password validations for Regular users and staff (i.e. Admins).
Also, make validations on the `is_staff` and `is_superuser` in case they were set as `False` for any reason.

You could use Registration Validations inside serializers (later on); @AhmedRamadanAR could implement them.

AAEmara commented on backend/accounts/models.py:
```diff
@@ -1,3 +1,103 @@
+    
+    email = models.EmailField(unique=True)
+    passwordhash = models.CharField(max_length=128)
+    full_name = models.CharField(max_length=255)
```
Kindly, break the `full_name` into:
  - `first_name`
  - `last_name`

AAEmara commented on backend/accounts/models.py:
```diff
@@ -1,3 +1,103 @@
+    ]
+    
+    email = models.EmailField(unique=True)
+    passwordhash = models.CharField(max_length=128)
```
Kindly, change the variable `passwordhash` into `password`, so it could be integrated correctly with `set_password` in `creat_user` method.

AAEmara commented on backend/accounts/models.py:
```diff
@@ -1,3 +1,103 @@
+        ('customer', 'Customer'),
+        ('seller', 'Seller'),
+        ('shipping_company', 'Shipping Company'),
+    ]
```
Kindly, remove ROLE_CHOICES, and replace it with a class named `Roles`, where we could use it's attribute through:
`Roles.choices`.

This would make the `User` class code much cleaner.

AAEmara commented on backend/accounts/models.py:
```diff
@@ -1,3 +1,103 @@
+    business_license = models.CharField(max_length=255)
+    tax_id = models.CharField(max_length=100)
+    verified = models.BooleanField(default=False)
+    account_status = models.CharField(max_length=20, choices=STATUS_CHOICES, default='pending')
```
Kindly, remove STATUS_CHOICES, and replace it with a class named Status, where we could use it's attribute through:
Status.choices.

AAEmara commented on backend/accounts/models.py:
```diff
@@ -1,3 +1,103 @@
+    is_default = models.BooleanField(default=False)
+
+    def __str__(self):
+        return f"{self.street}, {self.city}"
```
Move the `Address` model to the addresses app in order to achieve separation of concerns concept.

AAEmara commented on backend/accounts/admin.py:
```diff
@@ -1,3 +1,10 @@
 from django.contrib import admin
+from .models import *
```
Enhancement: Import individual models.

AAEmara commented on backend/accounts/admin.py:
```diff
@@ -1,3 +1,10 @@
+
+@admin.register(Address)
+class AddressAdmin(admin.ModelAdmin):
+    list_display = ('user', 'city', 'is_default')
```
Move this into the `addresses` django app.

AAEmara commented on backend/accounts/admin.py:
```diff
@@ -1,3 +1,10 @@
-# Register your models here.
+@admin.register(User)
+class UserAdmin(admin.ModelAdmin):
+    list_display = ('email', 'full_name', 'role', 'is_active')
```
Don't forget to divide `'full_name'` into `'first_name'` and `'last_name'`.

AAEmara commented on backend/core/settings.py:
```diff
@@ -40,6 +45,9 @@
     'django.contrib.staticfiles',
     'corsheaders',
     'rest_framework',
+    
```
Delete the extra line.

AAEmara commented on backend/core/settings.py:
```diff
@@ -13,9 +13,14 @@
 
+MEDIA_URL = '/media/'  # For product images
+MEDIA_ROOT = os.path.join(BASE_DIR, 'media')
+
```
Delete the extra line.

AAEmara reviewed (changes requested):
Great job @abdelsalam101! Just minor changes needed and we are set to merge this PR.

@AhmedRamadanAR, you could handle those changes before heading to the JWT feature.

AhmedRamadanAR committed:
feat(auth): update core user and address models

Updated models for the authentication and address modules, including:
- Refactored `accounts.models.User` to support custom user logic.
- Created initial migration `0002_initial.py` for user model updates.
- Registered updated models in Django admin.
- Created `addresses.models.Address` with essential fields.
- Added initial migration `0001_initial.py` for addresses.
- Registered address model in Django admin.

Closes #6

AAEmara reviewed (approved):
Great job @AhmedRamadanAR! Everything is alright 👍🏼
